torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -877,29 +883,38 @@ class AsyncSparseAllreduceWork : public ProcessGroupGloo::AsyncWork {
   std::vector<at::Tensor> allgather_indices(
       const at::Tensor& tensor,
       const std::vector<SparseTensorMetadata>& metadata) {
-    auto max_nnz = metadata[0].nnz();
-    for (size_t i = 1; i < metadata.size(); i++) {
-      max_nnz = std::max(max_nnz, metadata[i].nnz());
+    const auto sparseDim = tensor.sparse_dim();
+
+    std::vector<size_t> counts(context->size);
+    int64_t totalSize = 0;
+    for (size_t i = 0; i < metadata.size(); i++) {
+      counts[i] = metadata[i].nnz() * sparseDim;
+      totalSize += counts[i];
     }
 
-    // There are #sparse_dim() 1-dimensional tensors with nnz elems per rank.
-    auto buffer =
-        at::empty({context->size, tensor.sparse_dim(), max_nnz}, at::kLong);
-    buffer.select(0, context->rank)
-        .narrow(1, 0, tensor._nnz())
-        .copy_(tensor.indices());
+    auto output = at::empty({totalSize}, at::kLong);
 
-    // Allgather indices.
-    gloo::AllgatherOptions opts(context);
-    opts.setOutput(buffer.data_ptr<int64_t>(), buffer.numel());
+    // tensors copied from cuda may not be contiguous, get a contiguous
+    // tensor before use its data_ptr
+    auto input = tensor.indices().contiguous();
```

Comment:
The inputs are always coalesced before running the algorithm and I was thinking that that implies they'll be contiguous.

Is this not the case?